### PR TITLE
datapath: Share IPv4-or-IPv6 union between structures/maps

### LINKED ISSUE #23657: datapath: Share IPv4-or-IPv6 union between structures/maps
We use the following union in a couple BPF maps to encode both IPv4 and IPv6 addresses. We should refactor between the different maps that use that, to avoid redefining the union all over the datapath.
```c
union {
	struct {
		__u32 ip4;
		__u32 pad4;
		__u32 pad5;
		__u32 pad6;
	};
	union v6addr ip6;
};
```

### PR DESCRIPTION
**Pull request by EItanya**
Fixes: #23657

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

Fixes: #issue-number

### CODE REVIEW

**Comment by EItanya** [comment]
I would like to note that there are many more structs throughout the codebase that contain this type inlined, but I wanted to keep the scope of the PR down.

Also, as this is my first PR to Cilium I would love comments on the style of the C code!

File: USERS.md [hunk 1]
```diff
@@ -512,7 +512,7 @@ Users (Alphabetically)
     * N: Solo.io
       D: Cilium is part of Gloo Application Networking platform, with a “batteries included but swappable” manner
       U: CNI, Network Policies
-      Q: @linsun
+      Q: @linsun, @EItanya
```
Comment (gentoo-root) [inline USERS.md:515]: Could you make this change in a separate commit?

File: bpf/lib/ip_addr.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#include "common.h"
+
+union ip_address {
```
Comment (gentoo-root) [inline bpf/lib/ip_addr.h:3]: What's the purpose of this change? In other words, why is it useful to declare this union explicitly?

In my opinion, extracting the common code from ipcache_key and node_key doesn't bring much value by itself, especially considering that the field access is more verbose now, and other usages (e.g., endpoint_key, tunnel_key, tunnel_value) remain not converted. Is there a bigger goal that I don't see, and future plan on this refactoring?

We usually don't merge pull requests that are simply code churn without a clear impact. However, if you explain your end goal, this change might be a valid first step.

File: bpf/lib/ip_addr.h [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  struct {
+    __u32		ip4;
```
Comment (gentoo-root) [inline bpf/lib/ip_addr.h:5]: We use tabs for indentation, not two spaces.

File: bpf/lib/ip_addr.h [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    __u32		pad4;
+    __u32		pad5;
+    __u32		pad6;
+  };
+  union v6addr	ip6;
+} __packed;
```
Comment (gentoo-root) [inline bpf/lib/ip_addr.h:11]: Each line of the file, including the last one, should end with a newline character.

(There is usually a setting in text editors and IDEs that forces it.)

File: bpf/lib/ip_addr.h [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#include "common.h"
```
Comment (gentoo-root) [inline bpf/lib/ip_addr.h:1]: Why do you need a new file for a single union? Can it be defined in one of the existing header files that matches its meaning?

File: test/k8s/services.go [hunk 6]
```diff
@@ -645,7 +645,8 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`,
 		})
 
 		SkipContextIf(func() bool {
-			return helpers.DoesNotRunWithKubeProxyReplacement() || helpers.RunsOnAKS() || helpers.DoesNotExistNodeWithoutCilium()
+			return helpers.DoesNotRunWithKubeProxyReplacement() || helpers.RunsOnAKS() || helpers.DoesNotExistNodeWithoutCilium() ||
+				helpers.SkipQuarantined()
```
Comment (gentoo-root) [inline test/k8s/services.go]: This change doesn't look related to the code style change you are doing in this commit. Skipping the test looks suspicious.

**Review by gentoo-root — CHANGES_REQUESTED** [review]
Thanks for your first contribution!

I left some comments, including code style ones, as you asked.

One more note: to maintain cleaner commit history, we usually rebase the pull request branch instead of continuously merging main into that branch.

File: test/k8s/services.go (hunk 6 above)
Comment (EItanya) [reply test/k8s/services.go]: I think this may have come in from a bad merge with upstream, I can look into it

File: USERS.md (hunk 1 above)
Comment (EItanya) [reply USERS.md:515]: Yup!

File: bpf/lib/ip_addr.h (hunk 2 above)
Comment (EItanya) [reply bpf/lib/ip_addr.h:3]: To be honest I don't have a specific goal in making this change. I was looking for opportunities to contribute, and issue #23657 seemed like a good option. I interpreted that issue as wanting the current change, but I definitely could have misunderstood. What was the intention of that issue?  I see that @pchaigno created that issue, and liked this comment so maybe he can weigh in as well.

File: bpf/lib/ip_addr.h (hunk 5 above)
Comment (EItanya) [reply bpf/lib/ip_addr.h:1]: I put this in a different file because although the issue mentioned that this type should be combined within this file, it is used in many many other places around the code base.

File: bpf/lib/ip_addr.h (hunk 2 above)
Comment (pchaigno) [reply bpf/lib/ip_addr.h:3]: The intended goal of https://github.com/cilium/cilium/issues/23657 is to have a single definition of that union in the codebase. I think it's possible to do that without changing all users. I agree with Maxim that we should avoid making this a huge change that touches a lot of places or it's going to cause a lot of backporting conflicts and won't be worth it. We should also ensure to remove all such copies of the union definition and not do this piecemeal.

Comment (EItanya) [reply bpf/lib/ip_addr.h:3]: That makes sense to me. Maybe this isn’t such a good first issue then, the scope isn’t very clear.

Comment (pchaigno) [reply bpf/lib/ip_addr.h:3]: Can I help clarify anything? If you mean which unions to replace, then you can find them with `git grep union bpf/lib/`. They have different field names but you should be able to recognize the pattern.

Comment (EItanya) [reply bpf/lib/ip_addr.h:3]: Sure, I guess I'm a little confused by your previous comment. You mentioned not wanting to make it a huge change, but then also mentioned you don't want to do it piece meal. So I'm not sure what exactly the scope is here, and how you'd like to space it out.

As you mentioned, a similar `union` is defined elsewhere in the codebase, for instance: 
1. https://github.com/cilium/cilium/blob/1.13.2/bpf/lib/common.h#L260-L266
2. https://github.com/cilium/cilium/blob/1.13.2/bpf/lib/common.h#L245-L251

However, if I were to wholesale replace all of these uses with a new combined `union`, wouldn't that require a pretty large change as each access would change? I could definitely be missing a nuance that would allow me to do this without a large change to each use.

Sorry in advance for the confusion, and thanks for the help!

Comment (gentoo-root) [reply bpf/lib/ip_addr.h:3]: > You mentioned not wanting to make it a huge change, but then also mentioned you don't want to do it piece meal.

I think what Paul meant by a "huge change" is the side effect it has on various places all over the code: when you access the ip4/ip6 fields, you have to use ip.ip4/ip.ip6 now, and it's a lot of changes all over the codebase, which is likely to cause conflicts with backports. Paul said:

> I think it's possible to do that without changing all users.

So the idea here is to replace all unions at once (=not do it piecemeal), but minimize the impact on the _code that uses these fields_ (=avoid a huge change). I'm not sure it's possible, though: we are replacing an anonymous union with a named one, and we have to name the corresponding field (at the same time we probably don't want magic macros to work around it). @pchaigno?

Comment (EItanya) [reply bpf/lib/ip_addr.h:3]: Totally agree about the verbosity issues with this approach, and I personally don't love magic macros as a rule 😆. I can do some research to see if there's another way.

Comment (pchaigno) [reply bpf/lib/ip_addr.h:3]: We already use macros to avoid code duplication in C structures. See `__notify_common_hdr` and `__notify_pktcap_hdr`. I think it's possible to do the same here without magic. The difficulty is probably around the different names these map structures use for their IPv4-or-IPv6 union field? How much code churn would we end up with if we renamed to the most commonly-used fields?
